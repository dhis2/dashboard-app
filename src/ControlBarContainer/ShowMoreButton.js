import React from 'react';
import i18n from 'd2-i18n';
import { colors } from '../colors';

const ShowMoreButton = ({ onClick, isMaxHeight }) => (
    <div style={{ textAlign: 'center' }}>
        <div
            onClick={onClick}
            style={{
                paddingTop: 3,
                fontSize: 11,
                fontWeight: 700,
                color: colors.royalBlue,
                textTransform: 'uppercase',
                cursor: 'pointer',
                visibility: 'visible',
            }}
        >
            {isMaxHeight ? i18n.t('Show less') : i18n.t('Show more')}
        </div>
    </div>
);

export default ShowMoreButton;
