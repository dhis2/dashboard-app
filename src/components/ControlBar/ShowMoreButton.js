import React from 'react';
import i18n from '@dhis2/d2-i18n';
import { colors } from '@dhis2/ui-core';
import { withStyles } from '@material-ui/core/styles';

export const SHOWMORE_BAR_HEIGHT = 16;

const styles = {
    showMore: {
        color: colors.grey700,
        cursor: 'pointer',
        fontSize: 11,
        paddingTop: 5,
        '&:hover': {
            textDecoration: 'underline',
        },
    },
    disabled: {
        paddingTop: 5,
        color: colors.grey500,
        fontSize: 11,
        cursor: 'not-allowed',
    },
};

export const ShowMoreButton = ({ onClick, isMaxHeight, classes, disabled }) => {
    return (
        <div style={{ textAlign: 'center' }}>
            {disabled ? (
                <div className={classes.disabled}>{i18n.t('Show more')}</div>
            ) : (
                <div className={classes.showMore} onClick={onClick}>
                    {isMaxHeight ? i18n.t('Show less') : i18n.t('Show more')}
                </div>
            )}
        </div>
    );
};

export default withStyles(styles)(ShowMoreButton);
