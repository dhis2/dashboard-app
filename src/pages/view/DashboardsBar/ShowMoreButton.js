import i18n from '@dhis2/d2-i18n'
import { Tooltip } from '@dhis2/ui'
import PropTypes from 'prop-types'
import React, { useRef } from 'react'
import { ChevronDown, ChevronUp } from './assets/icons'
import classes from './styles/ShowMoreButton.module.css'

const ShowMoreButton = ({ onClick, dashboardBarIsExpanded, disabled }) => {
    const containerRef = useRef(null)
    const buttonLabel = dashboardBarIsExpanded
        ? i18n.t('Show fewer dashboards')
        : i18n.t('Show more dashboards')

    const onButtonClicked = () => {
        // The click may happen on the svg or path
        // element of the button icon.
        // In that case it is necessary to trigger
        // the mouseout on the button element
        // to ensure that the tooltip is removed
        const buttonEl = containerRef.current.children[0]
        const event = new MouseEvent('mouseout', {
            bubbles: true,
            cancelable: false,
        })

        onClick()
        buttonEl.dispatchEvent(event)
    }

    return (
        <div className={classes.container} ref={containerRef}>
            {disabled ? (
                <div className={classes.disabled}>
                    <ChevronDown />
                </div>
            ) : (
                <Tooltip content={buttonLabel} placement="bottom">
                    {({ onMouseOver, onMouseOut, ref }) => (
                        <button
                            className={classes.showMore}
                            onClick={onButtonClicked}
                            data-test="showmore-button"
                            aria-label={buttonLabel}
                            ref={ref}
                            onMouseOver={onMouseOver}
                            onMouseOut={onMouseOut}
                        >
                            {dashboardBarIsExpanded ? (
                                <ChevronUp />
                            ) : (
                                <ChevronDown />
                            )}
                        </button>
                    )}
                </Tooltip>
            )}
        </div>
    )
}

ShowMoreButton.propTypes = {
    dashboardBarIsExpanded: PropTypes.bool,
    disabled: PropTypes.bool,
    onClick: PropTypes.func,
}

export default ShowMoreButton
