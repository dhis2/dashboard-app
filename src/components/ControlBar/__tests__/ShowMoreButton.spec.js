import React from 'react';
import { shallow } from 'enzyme';
import { ShowMoreButton } from '../ShowMoreButton';

describe('ShowMoreButton', () => {
    it('renders correctly when at maxHeight', () => {
        const button = shallow(
            <ShowMoreButton
                onClick={() => {}}
                isMaxHeight={true}
                classes={{ showMore: {} }}
            />
        );
        expect(button).toMatchSnapshot();
    });

    it('renders correctly when not at maxHeight', () => {
        const tree = shallow(
            <ShowMoreButton
                onClick={() => {}}
                isMaxHeight={false}
                classes={{ showMore: {} }}
            />
        );

        expect(tree).toMatchSnapshot();
    });

    describe('actions', () => {
        let props;
        let shallowShowMoreButton;
        const showMoreButton = () => {
            if (!shallowShowMoreButton) {
                shallowShowMoreButton = shallow(<ShowMoreButton {...props} />);
            }
            return shallowShowMoreButton;
        };

        beforeEach(() => {
            props = {
                onClick: undefined,
                isMaxHeight: undefined,
                classes: { showMore: {} },
            };
            shallowShowMoreButton = undefined;
        });

        it('changes text when isMaxHeight is changed', () => {
            props.isMaxHeight = true;
            const button1Text = showMoreButton().text();

            shallowShowMoreButton = undefined;
            props.isMaxHeight = false;
            const button2Text = showMoreButton().text();

            expect(button1Text.length).toBeGreaterThan(0);
            expect(button2Text.length).toBeGreaterThan(0);
            expect(button1Text).not.toEqual(button2Text);
        });

        describe('when onClick is defined', () => {
            beforeEach(() => {
                props.onClick = jest.fn();
            });

            it('triggers onClick when div clicked', () => {
                showMoreButton()
                    .childAt(0)
                    .simulate('click');
                expect(props.onClick).toHaveBeenCalled();
            });
        });
    });
});
