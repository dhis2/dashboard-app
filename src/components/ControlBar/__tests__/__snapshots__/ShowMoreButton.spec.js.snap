// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowMoreButton renders correctly when at maxHeight 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ShowMoreButton
    classes={
      Object {
        "showMore": Object {},
      }
    }
    isMaxHeight={true}
    onClick={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className={Object {}}
        onClick={[Function]}
      >
        Show less
      </div>,
      "style": Object {
        "textAlign": "center",
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Show less",
        "className": Object {},
        "onClick": [Function],
      },
      "ref": null,
      "rendered": "Show less",
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className={Object {}}
          onClick={[Function]}
        >
          Show less
        </div>,
        "style": Object {
          "textAlign": "center",
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Show less",
          "className": Object {},
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Show less",
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`ShowMoreButton renders correctly when not at maxHeight 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ShowMoreButton
    classes={
      Object {
        "showMore": Object {},
      }
    }
    isMaxHeight={false}
    onClick={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className={Object {}}
        onClick={[Function]}
      >
        Show more
      </div>,
      "style": Object {
        "textAlign": "center",
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Show more",
        "className": Object {},
        "onClick": [Function],
      },
      "ref": null,
      "rendered": "Show more",
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className={Object {}}
          onClick={[Function]}
        >
          Show more
        </div>,
        "style": Object {
          "textAlign": "center",
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Show more",
          "className": Object {},
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Show more",
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
